### ISSUE TITLE
`assertObjectMatch` throws error for non-object

### ISSUE THREAD

**Issue by vwkd** [denoland/deno_std#3451]
**Describe the bug**

`assertObjectMatch` should check partial equality of two objects.

Currently, when calling `assertObjectMatch` with a non-object in the first argument, it throws a cryptic `Uncaught TypeError: Invalid value used as weak map key` error and if it's the second argument `Uncaught TypeError: Cannot use 'in' operator to search for 'yyy' in xxx`.

Confusing cryptic errors depending on argument order aside, for top-level non-object inputs we should use `assertEquals`. But the above errors are also thrown for any nested property due to the use of recursion.

For example, currently `assertObjectMatch` breaks if some deep property that expects a nested object turns out to be `null` or `undefined`. This should pass, for example to check a partially successful GraphQL response.

**Steps to Reproduce**

```js
import { assertObjectMatch } from "https://deno.land/std@0.191.0/testing/asserts.ts";

assertObjectMatch({ foo: null }, { foo: { bar: 42 }});
// error: Uncaught TypeError: Invalid value used as weak map key

assertObjectMatch({ foo: { bar: 42 }}, { foo: null });
// switch arguments and it works

assertObjectMatch(null, { foo: 42 });
// error: Uncaught TypeError: Invalid value used as weak map key

assertObjectMatch({ foo: 42 }, null);
// error: Uncaught TypeError: Cannot use 'in' operator to search for 'foo' in null

assertObjectMatch(21, 42);
// error: Uncaught TypeError: Invalid value used as weak map key
```

**Expected behavior**

`assertObjectMatch` should handle non-object inputs. It should pass a superset of inputs that pass `assertEquals`, meaning, it should throw less often, not more often.

**Environment**

- OS: macOS 12.6.6
- deno version: deno 1.34.1
- std version: 0.191.0

**Comment by luk3skyw4lker** [comment]
I'm taking this issue!

**Comment by luk3skyw4lker** [comment]
Just an update: I'm still working on it, but I'm having a little trouble in understanding the code behind the function, I'll be sending a PR anytime soon

**Comment by JakeAve** [comment]
Hey @luk3skyw4lker I'm not sure what solutions you've come up with, but I slightly disagree with putting the primitive types as assertion errors. Everything in JS is an object of sorts, but we kind of all understand what we mean by assertObjectMatch. I thought it made sense to have primitives throw a TypeError rather than an Assertion error because honestly, you should use assertEquals for primitives. Also, the issue with null is the infamous `typeof null === 'object` bug. What suggestions do you have? Do you want to branch off mine or make a different PR?

**Comment by JakeAve** [comment]
You were probably closer than you think. The whole thing is just iterating over weakmaps until in arrives on a single primitive type, but since it thought `null` is an object with properties, it was trying to iterate over `null`. So on line 576, you make sure it's not null `(typeof value === "object" && value !== null)` , and if it is null send it to line 600, with other primitives like strings, numbers, instead of sending it through more recursion on line 596. But this isn't your fault, it's a notorious JS bug that is too old to easily fix where `typeof null === 'object'`.

**Comment by kt3k** [comment]
should be fixed by #3468
